File: src/algorithms/python/Kth_Smallest_Element_in_a_BST.py
# -*- coding: utf-8 -*-


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def pre_dfs(self, node):
        _left = self.pre_dfs(node.left) if node.left else 0
        node.left_cnt = _left + 1
        _right = self.pre_dfs(node.right) if node.right else 0
        return _left + _right + 1

    # @param {TreeNode} root
    # @param {integer} k
    # @return {integer}
    def kthSmallest(self, root, k):
        self.pre_dfs(root)
        node = root
        while True:
            if node.left_cnt < k:
                k -= node.left_cnt
                node = node.right
            elif node.left_cnt > k:
                node = node.left
            else:
                return node.val


if __name__ == "__main__":
    solution = Solution()
    l = [2, 1, 3]
    l = map(TreeNode, l)
    l[0].right, l[0].left = l[1], l[2]
    print solution.kthSmallest(l[0], 3)

File: src/algorithms/cpp/Kth_Smallest_Element_in_a_BST.cpp
#include <iostream>
#include <cstdio>
#include <cstring>
#include <cmath>
#include <cstdlib>
#include <algorithm>
#include <vector>
#include <map>
#include <set>
using namespace std;
#define out(v) cerr << #v << ": " << (v) << endl
#define SZ(v) ((int)(v).size())
const int maxint = -1u>>1;
template <class T> bool get_max(T& a, const T &b) {return b > a? a = b, 1: 0;}
template <class T> bool get_min(T& a, const T &b) {return b < a? a = b, 1: 0;}

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
private:
    int cnt, result;

private:
    void dfs(TreeNode *root) {
        if (root->left) {
            if (cnt > 0)    dfs(root->left);
        }
        cnt --;
        if (cnt == 0)   result = root->val;
        if (root->right) {
            if (cnt > 0)    dfs(root->right);
        }
    }

public:
    int kthSmallest(TreeNode* root, int k) {
        this->cnt = k;
        dfs(root);
        return result;
    }
};

int main() {
    return 0;
}
